### PR TITLE
feat: offline behavior-token motion-prior diagnostics prototype (issue #4627)

### LINKED ISSUE #4627: Experiment: behavior-token motion priors for scenario interaction traces
## Context

The arXiv paper `2606.29148` motivates a useful but speculative research direction: pretraining reusable discrete motion/control priors from large motion corpora, then adapting them to downstream tasks instead of learning every behavior from scratch.

For `robot_sf_ll7`, this should **not** become a near-term roadmap item. It is a low-priority experimental note to preserve the idea for later exploration if the existing scenario, pedestrian-interaction, and RL infrastructure becomes mature enough.

## Research question

Can pedestrian/robot interaction traces in `robot_sf_ll7` be represented as short discrete **behavior tokens** or latent skill codes, and can those tokens be used as a diagnostic prior for scenario grouping, failure analysis, or downstream RL initialization?

## Possible experimental slice

Start with an offline, non-invasive prototype:

1. Collect existing simulation traces from representative scenarios.
2. Encode short trajectory windows into compact descriptors, for example:
   - relative robot-pedestrian geometry,
   - velocity and acceleration changes,
   - TTC margin evolution,
   - clearance evolution,
   - path blocking / yielding / passing patterns,
   - recovery smoothness after near-conflict.
3. Cluster or quantize these windows into a small discrete vocabulary.
4. Inspect whether the resulting tokens correspond to interpretable interaction motifs:
   - yielding,
   - assertive passing,
   - oscillatory negotiation,
   - emergency braking,
   - pedestrian displacement,
   - deadlock / freeze,
   - unsafe cut-in.
5. Use the tokens only for diagnostics at first:
   - scenario similarity,
   - collision-cause grouping,
   - regression-test stratification,
   - coverage reporting.

## Explicit non-goals

- Do not train a new controller yet.
- Do not replace the current safety metrics.
- Do not introduce a large transformer dependency.
- Do not make this part of the main evaluation pipeline.
- Do not claim foundation-model behavior unless supported by evidence.

## Suggested implementation boundary

A minimal future prototype could live behind an experimental namespace, for example:

```text
experiments/behavior_tokens/
```

Possible files:

```text
experiments/behavior_tokens/extract_windows.py
experiments/behavior_tokens/quantize_trace_windows.py
experiments/behavior_tokens/inspect_token_motifs.py
experiments/behavior_tokens/README.md
```

The first version should operate only on saved traces and produce static artifacts such as CSV/JSON summaries and simple plots.

## Acceptance criteria for a first exploratory pass

- [ ] A small offline script can load saved interaction traces.
- [ ] Trajectory windows are converted into documented feature vectors.
- [ ] A simple clustering or quantization method assigns each window to a discrete token ID.
- [ ] The resulting token distribution can be summarized per scenario.
- [ ] At least 5–10 example windows per frequent token can be inspected manually.
- [ ] The README states that this is experimental, low priority, and not part of the validated metric stack.

## Priority

Very low. This should only be considered after the current dissertation-critical scenario, metric, prediction, and evaluation issues are stable.

## References

- arXiv:2606.29148 — generative pretrained controllers / reusable motion priors concept.

### PR DESCRIPTION
**Pull request by ll7**
## Reviewer-critical summary

- **What changed:** New offline, experimental `experiments/behavior_tokens/` namespace that turns saved benchmark interaction traces into discrete "behavior tokens" for diagnostics (3 standalone CLIs + shared schemas + tests), implementing the maintainer plan on #4627.
- **Why now:** #4627 is auto-admitted from the ready queue; the plan is fully specified and self-contained (offline, CPU-only), so the complete first exploratory slice is achievable in one session.
- **Claim boundary:** Very low priority, exploratory, **diagnostic-only** tooling. Behavior tokens are heuristic descriptors of saved traces — **not** validated metrics, benchmark evidence, a release gate, or paper/dissertation claim support. **No safety decision may depend on the tokens.** No new controller training, no transformer dependency, no benchmark-pipeline integration.
- **Required validation:** ruff check/format, `tests/experiments/test_behavior_tokens.py`, and a 3-stage end-to-end smoke on a synthetic trace file (results below).
- **Remaining blocker:** None for this first exploratory pass. Running the pipeline on *real* saved campaign traces (no local `episodes.jsonl` with `simulation_step_trace` exists in this checkout) and any downstream diagnostic use remain future, opt-in work.

Closes #4627

## Contract delta

- **New experimental namespace** `experiments/behavior_tokens/` (offline, read-only):
  - `extract_windows.py` — slides fixed windows over `algorithm_metadata.simulation_step_trace.steps` → documented feature vectors (clearance min/mean/slope, TTC proxy min/slope, robot-speed stats, accel/command-change proxy, stop/yield, oscillation, near-conflict recovery, relative/ped speed). Non-derivable features are JSON `null` + listed in `missing_features` (never fabricated zeros); rows without a trace are **skipped with an explicit reason**.
  - `quantize_trace_windows.py` — standardizes finite feature columns and assigns each valid window a **deterministic** discrete token id via k-means (scikit-learn, NumPy-only fallback). Token ids are **canonicalized by cluster center** so repeated runs and both backends agree. **Fails closed** (raises, non-zero exit) when no finite columns or no valid windows remain, rather than emitting empty/fabricated assignments.
  - `inspect_token_motifs.py` — token summary by scenario/planner/outcome/row-status, heuristic `*_candidate` motif labels, and **bounded** example windows per frequent token.
  - `schemas.py` — shared `FEATURE_NAMES` vocabulary + schema-version constants embedded in every artifact for provenance.
- **New schema-version strings** (artifact provenance): `behavior-token-features.v1`, `behavior-token-window.v1`, `behavior-token-quantizer.v1`, `behavior-token-inspection.v1`.
- **No new fail-closed benchmark blocker** and **no change to any existing metric/benchmark/schema surface.** All outputs default under git-ignored `output/experiments/behavior_tokens/...`.
- **Compatibility:** additive only. `pyproject.toml` gains an `experiments/**` ruff per-file-ignore mirroring the existing one-off-script allowances (`T201`, `PLC0415`, `TC001`, `TC002`, `D417`, `DOC201`).

## Validation results

| Command | Result |
| --- | --- |
| `ruff check experiments/behavior_tokens tests/experiments/test_behavior_tokens.py` | **All checks passed!** |
| `ruff format --check` (same paths) | reformatted then clean |
| `pytest tests/experiments/test_behavior_tokens.py -q` | **7 passed** |
| `python .../extract_windows.py --help` / `quantize_... --help` / `inspect_... --help` | exit 0 (all three) |
| 3-stage end-to-end smoke on a synthetic 25-row `episodes.jsonl` | 96 windows extracted, **1 no-trace row skipped** (`no_simulation_step_trace`), quantized → 8 tokens (`sklearn.KMeans`, 0 excluded), inspection wrote summary + `frequent_tokens.md` + bounded (≤5) example files |
| `scripts/tools/sync_ai_config.py --check` | AI config links validated: 7 symlinks (no drift) |

Tests cover: expected window count; missing-optional-field null discipline (no fabricated zeros) with a genuine-zero counter-check; no-trace / too-short skip reasons; quantizer determinism; one token id per valid window; bounded example selection; and the experimental claim boundary in README + generated metadata.

## Out of scope (explicit)

- **No full benchmark campaign run.** No Slurm/GPU submission. No paper/dissertation claim edits.
- No controller training, no transformer/model dependency, no benchmark/reporting-pipeline integration, no metric-stack change.
- No run on real saved campaign traces (none present locally with `simulation_step_trace`); the smoke used a synthetic trace file.

## State propagation

Low-churn issue (only the maintainer plan comment before this PR). Durable state surface updated via `CHANGELOG.md` (`[Unreleased] → Added`). This PR is schema/behavior only; no transient queue-routing state is encoded in tracked files.

<details>
<summary>Governance appendix</summary>

- **Research Result Guidance:** `NA — exploratory diagnostic tooling.` No claim/hypothesis is asserted; token ids and motif labels are heuristic candidates for manual inspection. Target synthesis surface if pursued: a future diagnostic context note under `docs/context/`.
- **Downstream Propagation:** parent issue #4627 (this PR); CHANGELOG updated. No claim map, leaderboard, registry, or benchmark report is affected (diagnostic-only). Disposable outputs live under git-ignored `output/experiments/behavior_tokens/`.
- **Follow-up:** first use on durable/real traces + any diagnostic application (scenario similarity, collision-cause grouping, regression stratification) is deferred, opt-in work; not required for this first exploratory slice.
- **Freshness:** re-checked issue #4627 comments immediately before opening — only the original maintainer plan comment (2026-07-06T14:00:34Z) exists; no newer guidance; no duplicate/overlapping PR.
</details>

### CODE REVIEW

**Comment by ll7** [comment]
## Gate review — PR #4657

**Phase 1 (review):** read the full diff (1886 add / 0 del): new isolated `experiments/behavior_tokens/` namespace (3 CLIs + schemas + tests), pyproject lint scope, CHANGELOG.

Findings:
1. **No correctness findings.** Quantization is deterministic and fail-closed: `_numpy_kmeans` seeds `np.random.default_rng(seed)`, scikit-learn path uses `random_state=seed`, and both are canonicalized by cluster center (`_canonicalize_labels` sorts by rounded center) so backends/reruns agree. `run_quantization` raises `ValueError` (non-zero exit) when no finite feature columns or no fully-finite windows exist — no silent empty output. `extract_windows` emits JSON `null` + `missing_features` for non-derivable features instead of fabricating zeros, and skips trace-less rows with an explicit reason.
2. **No runtime/benchmark-semantics change.** Only shared-surface edit is a Ruff per-file-ignore for `experiments/**/*.py` (mirrors the existing `scripts/**` allowances) — no new install dependency; scikit-learn use has a NumPy-only fallback. No planner/wrapper/metric/scenario path touched; strictly offline, CPU-only, read-only diagnostic tooling with an explicit "no safety decision may depend on the tokens" boundary.

**Open review threads: 0** (gemini + CodeRabbit both rate-limited — no substantive threads).

**Phase 2 (fix):** nothing in-scope to fix.

**Phase 3 (merge):** `Closes #4627` is warranted — all six "first exploratory pass" acceptance criteria map to delivered CLIs (load traces → documented feature vectors → discrete token IDs → per-scenario token summary → ≥5–10 inspectable windows via `inspect_token_motifs.py` → README experimental/not-validated boundary); the maintainer implementation plan matches the delivered namespace exactly. Running on real campaign traces is out-of-scope future work per the issue. 16/16 checks green, MERGE=CLEAN. Merging.
